backend/src/database/services/base-service.ts

```diff
@@ -0,0 +1,172 @@
+import type { UUID } from 'digital-fuesim-manv-shared';
+import type {
+    DataSource,
+    DeleteResult,
+    EntityManager,
+    EntityTarget,
+    FindManyOptions,
+    FindOneOptions,
+    FindOptionsWhere,
+} from 'typeorm';
+import { EntityNotFoundError } from 'typeorm';
+import { DatabaseError } from '../database-error';
+import type { BaseEntity } from '../entities/base-entity';
+
+export abstract class BaseService<
+    Entity extends BaseEntity<Entity, any>,
+    Creatable,
+    Updatable
+> {
+    public constructor(protected readonly dataSource: DataSource) {}
+
+    protected abstract getCreateEntityObject(
+        dto: Creatable,
+        manager: EntityManager
+    ): Entity | Promise<Entity>;
+
+    protected abstract getUpdateEntityObject(
+        initialObject: Entity,
+        dto: Updatable,
+        manager: EntityManager
+    ): Entity | Promise<Entity>;
+
+    protected abstract readonly entityTarget: EntityTarget<Entity>;
+
+    public async create(
+        creator: Creatable,
+        entityManager?: EntityManager
+    ): Promise<Entity> {
+        const create = async (manager: EntityManager) =>
+            manager.save(await this.getCreateEntityObject(creator, manager));
+        return entityManager
+            ? create(entityManager)
+            : this.dataSource.transaction(create);
+    }
+
+    public async findAll(
+        options?: FindManyOptions<Entity>,
+        entityManager?: EntityManager
+    ): Promise<Entity[]> {
+        const find = async (manager: EntityManager) =>
+            manager.find(this.entityTarget, options);
+        return entityManager
+            ? find(entityManager)
+            : this.dataSource.transaction(find);
+    }
+
+    /**
+     * Finds the first row matching the {@link options}.
+     *
+     * @returns The found row, or `null` if no matching row was found and {@link mustExist} is `false`.
+     * @throws {@link DatabaseError} when {@link mustExist} is `true` and no row has been found.
+     */
+    public async findOne<TExists extends false | true>(
```

Comment:
Or two separate functions: `findOne(): Promise<Entity | null>` and `getOne(): Promise<Entity>`. `getOne` would just call findOne and add the extra catch.

Reply:
> Is this really necessary? I would probably remove the `mustExist` and the generic and just add a `!` if I know it must exist.

For future reference, this would not be what I want. `mustExist` represents a run-time assertion that the entity exists. Using `!` doesn't achieve that, and errors depending on this would happen later on with a much more obscure message than with the current solution.

(Apart from that I'll refactor this with the `findOne`/`getOne` solution described below.)